File: Cpp/src/SeqIdAndSeq.cpp
/*
 * SeqIdAndSeq.cpp
 *
 *  Created on: Jan 28, 2023
 *      Author: Adam Yongxin Ye @ BCH
 */

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>

#include "SeqIdAndSeq.h"
#include "YyxUtils.h"

void write_fasta(std::ostream *out, std::vector<SeqIdAndSeq> id_seqs){
	for(unsigned int k=0; k<id_seqs.size(); k++){
		SeqIdAndSeq id_seq = id_seqs[k];
//	for(SeqIdAndSeq id_seq : id_seqs){
		*out << ">" << id_seq.first << std::endl;
		*out << id_seq.second << std::endl;
	}
}

// reference: https://www.techiedelight.com/trim-string-cpp-remove-leading-trailing-spaces/
std::string chomp(const std::string &s)
{
    size_t end = s.find_last_not_of("\n\r");
    return (end == std::string::npos) ? "" : s.substr(0, end + 1);
}

std::vector<SeqIdAndSeq> read_fasta(std::string input_path){
	std::vector<SeqIdAndSeq> id_seqs;
	std::cerr << "[STEP] opening fasta file '" << input_path << "' ..." << std::endl;
	std::string now_name = "";
	std::string now_seq = "";
	// reference: https://stackoverflow.com/questions/7868936/read-file-line-by-line-using-ifstream-in-c
	std::string line;
	std::ifstream infile(input_path);
	if(! infile.good()){
		std::cerr << "Error: cannot open '" << input_path << "' for input" << std::endl;
		return id_seqs;
	}
	while(std::getline(infile, line)){
		line = chomp(line);
		if(line[0] == '>'){
			if(now_name != ""){
				// reference: https://stackoverflow.com/questions/14265581/parse-split-a-string-in-c-using-string-delimiter-standard-c
				now_name = now_name.substr(0, now_name.find(" "));
				id_seqs.push_back(std::make_pair(now_name, now_seq));
			}
			now_name = line.substr(1);
			now_seq = "";
		}else{
			// reference: https://stackoverflow.com/questions/735204/convert-a-string-in-c-to-upper-case
			std::transform(line.begin(), line.end(), line.begin(), ::toupper);
			now_seq += line;
		}
	}
	if(now_name != ""){
		// reference: https://stackoverflow.com/questions/14265581/parse-split-a-string-in-c-using-string-delimiter-standard-c
		now_name = now_name.substr(0, now_name.find(" "));
		id_seqs.push_back(std::make_pair(now_name, now_seq));
	}
	return id_seqs;
}

std::vector<SeqIdAndSeq> read_fastq(std::string input_path, int min_baseQ, int baseQ_coding_shift){
	std::vector<SeqIdAndSeq> id_seqs;
	std::cerr << "[STEP] opening fastq file '" << input_path << "' ..." << std::endl;
	std::string now_name = "";
	std::string now_seq = "";
	// reference: https://stackoverflow.com/questions/7868936/read-file-line-by-line-using-ifstream-in-c
	std::string lines[4];
	unsigned long NR = 1;
	std::ifstream infile(input_path);
	if(! infile.good()){
		std::cerr << "Error: cannot open '" << input_path << "' for input" << std::endl;
		return id_seqs;
	}
	while(std::getline(infile, lines[0]) && std::getline(infile, lines[1])){
		std::getline(infile, lines[2]);
		std::getline(infile, lines[3]);

		lines[0] = chomp(lines[0]);
		lines[1] = chomp(lines[1]);
		lines[3] = chomp(lines[3]);
		if(lines[0][0] == '@'){
			now_name = lines[0].substr(1);
			now_seq = lines[1];

			if(min_baseQ > 0){
				// convert bases with lower baseQ to Ns
				if(lines[1].length() != lines[3].length()){
					std::cerr << "Warning: invalid fastq format: different length of bases and baseQs at Line " << (NR+1) << " and " << (NR+3) << std::endl;
				}
				unsigned int len = lines[1].length();
				if(lines[3].length() < len){
					len = lines[3].length();
				}
				for(unsigned int i=0; i<len; i++){
					if(lines[3][i] + baseQ_coding_shift < min_baseQ){
						now_seq[i] = 'N';
					}
				}
			}

			id_seqs.push_back(std::make_pair(now_name, now_seq));
		}else{
			std::cerr << "Warning: invalid fastq format: first line not start with @ for Line " << NR << std::endl;
		}
		NR += 4;
	}
	return id_seqs;
}

File: Java/src/edu/harvard/hms/triededup/utils/SeqIdAndSeq.java
package edu.harvard.hms.triededup.utils;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

import htsjdk.samtools.SAMRecord;
import htsjdk.samtools.SamReader;
import htsjdk.samtools.SamReaderFactory;
import htsjdk.samtools.ValidationStringency;

/**
 * A structure class to store one pair of sequence ID and its sequence
 * 
 * @author Adam Yongxin Ye @ BCH
 *
 */
public class SeqIdAndSeq {
	public String seqId;
	public String sequence;
	
	public SeqIdAndSeq(String seqId, String sequence){
		this.seqId = seqId;
		this.sequence = sequence;
	}
	
	
	public static void write_fasta(PrintStream out, List<SeqIdAndSeq> id_seqs){
		for(SeqIdAndSeq id_seq : id_seqs){
			out.print('>');
			out.println(id_seq.seqId);
			out.println(id_seq.sequence);
		}
	}
	
	
	public static List<SeqIdAndSeq> read_fasta(String input_path) throws IOException{
		List<SeqIdAndSeq> id_seqs = new ArrayList<SeqIdAndSeq>();
		System.err.println(String.format("[STEP] opening fasta file %s ...", input_path));
		BufferedReader reader = new BufferedReader(new FileReader(input_path));
		String now_name = "";
		StringBuilder now_seq = new StringBuilder();
		String line;
		while((line = reader.readLine()) != null){
			line = line.trim();   // may change to stripTrailing after Java 11
			if(line.charAt(0) == '>'){
				if(now_name != ""){
					now_name = now_name.split(" ")[0];
					id_seqs.add(new SeqIdAndSeq(now_name, now_seq.toString()));
				}
				now_name = line.substring(1);
				now_seq = new StringBuilder();
			}else{
				now_seq.append(line.toUpperCase());
			}
		}
		if(now_name != ""){
			id_seqs.add(new SeqIdAndSeq(now_name, now_seq.toString()));
		}
		reader.close();
		
		return id_seqs;
	}
	
	public static List<SeqIdAndSeq> read_fastq(String input_path, int min_baseQ, int baseQ_coding_shift) throws IOException{
		List<SeqIdAndSeq> id_seqs = new ArrayList<SeqIdAndSeq>();
		System.err.println(String.format("[STEP] opening fastq file %s ...", input_path));
		BufferedReader reader = new BufferedReader(new FileReader(input_path));
		String now_name = "";
		char[] now_seq;
		String[] lines = new String[4];
		int NR = 1;
		while((lines[0] = reader.readLine()) != null){
			lines[1] = reader.readLine();
			lines[2] = reader.readLine();
			lines[3] = reader.readLine();
			if(lines[1] == null){
				break;
			}
			lines[0] = lines[0].trim();   // may change to stripTrailing after Java 11
			lines[1] = lines[1].trim();
			lines[3] = lines[3].trim();
			
			if(lines[0].charAt(0) == '@'){
				now_name = lines[0].substring(1);
				now_seq = lines[1].toCharArray();
				
				if(min_baseQ > 0){
					// convert bases with lower baseQ to Ns
					if(lines[1].length() != lines[3].length()){
						System.err.println(String.format("Warning: invalid fastq format: different length of bases and baseQs at Line %d and %d", (NR+1), (NR+3)));
					}
					int len = lines[1].length();
					if(lines[3].length() < len){
						len = lines[3].length();
					}
					for(int i=0; i<len; i++){
						if(lines[3].charAt(i) + baseQ_coding_shift < min_baseQ){
							now_seq[i] = 'N';
						}
					}
				}
				
				id_seqs.add(new SeqIdAndSeq(now_name, new String(now_seq)));
			}else{
				System.err.println(String.format("Warning: invalid fastq format: first line not start with @ for Line %d", NR));
			}
			NR += 4;
		}
		reader.close();
		
		return id_seqs;
	}
	
	public static List<SeqIdAndSeq> read_bam(String input_path, int min_baseQ, int baseQ_coding_shift) throws IOException{
		List<SeqIdAndSeq> id_seqs = new ArrayList<SeqIdAndSeq>();
		System.err.println(String.format("[STEP] opening sam/bam/cram file %s ...", input_path));
		// htsjdk   ref: https://github.com/samtools/htsjdk/blob/master/src/main/java/htsjdk/samtools/example/ExampleSamUsage.java
		final SamReader reader = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.LENIENT).open(new File(input_path));
		int NR = 1;
		for(final SAMRecord samRecord : reader){
			char[] now_seq = samRecord.getReadString().toCharArray();
			byte[] now_baseQs = samRecord.getBaseQualities();

			if(min_baseQ > 0){
				// convert bases with lower baseQ to Ns
				if(now_seq.length != now_baseQs.length){
					System.err.println(String.format("Warning: invalid sam/bam/cram format: different length of bases and baseQs at Line %d", NR));
				}
				int len = now_seq.length;
				if(now_baseQs.length < len){
					len = now_baseQs.length;
				}
				for(int i=0; i<len; i++){
					if(now_baseQs[i] + baseQ_coding_shift < min_baseQ){
						now_seq[i] = 'N';
					}
				}
			}
			
			id_seqs.add(new SeqIdAndSeq(samRecord.getReadName(), new String(now_seq)));
			
			NR++;
		}
		reader.close();
		
		return id_seqs;
	}
	
	
}
